### ISSUE TITLE
Actions are lost after END is dispatched

### ISSUE THREAD

**Issue by crizant** [bmealhouse/next-redux-saga#6]
Hi, I'm using `next-redux-saga` in sync mode, and I have an issue:
I see that in [`index.js`](https://github.com/bmealhouse/next-redux-saga/blob/master/index.js) line 21, it waits the saga tasks to be done, and restarts saga after that.
So that the actions dispatched between `END` is dispatched and the restart of saga, will not be processed by saga.

Is there any way to store those actions, and pass them to the saga middleware after the saga is restarted?

**Comment by bmealhouse** [comment]
Hi @crizant, good question. I'm not sure if this is possible. Can you share a code example that reproduces the issue?

Also, if you come up with a solution to the problem, feel free to submit a PR.

**Comment by crizant** [comment]
@bmealhouse Here it is:
https://github.com/crizant/next-redux-saga-issue

I am thinking about patching the `dispatch` function of the store, so that when it sees the `END` action,
Stop calling the real `dispatch` and store the subsequent actions in a queue.
Then after all saga task are done, dispatch the queued actions.

What do you think of my approach?

**Comment by crizant** [comment]
Alternatively, could this problem by solved by using action channels?
But after reading all documentation I still have no clue, I am just a newbie on `redux-saga`. =[

**Comment by bmealhouse** [comment]
> I am thinking about patching the dispatch function of the store, so that when it sees the END action,
stop calling the real dispatch and put the subsequent actions in a queue. Then after all saga tasks are done, dispatch another action (i.e. @@redux-saga/TasksDone) in next-redux-saga. When it sees the action, call the real dispatch to dispatch the queued actions.

This is an interesting idea. You may be able to accomplish the same functionality in a redux middleware. Then you wouldn't have to worry about patching `dispatch`.

Channels where the first thing that crossed my mind after reading your issue, but I wanted to see your example before I suggested it (they can get pretty ugly and difficult to understand).

You could also turn on `async` mode for this part of the app. Then `END` will never be dispatched on the client. If you took this approach, I would also recommend creating new UX patterns for content that is loading asynchronously.

If I can find some time, I will try to create some examples for you.

**Comment by crizant** [comment]
I can’t use async mode, since my application is like a blog, when using async mode and navigate from one article to another, the componentDidMount and routeChangeComplate callbacks will be executed immediately.
But I have other actions, like other api calls, google analytics etc, should be executed after the page data is loaded and dom is updated.
Therefore sync mode is my only option.

I will take a look at the possibility of using a custom middleware to solve the issue.

**Comment by bmealhouse** [comment]
Check out my solution using channels. As you can see it's really convoluted. 
[https://github.com/bmealhouse/next-redux-saga-issue/tree/redux-saga-channels](https://github.com/bmealhouse/next-redux-saga-issue/tree/redux-saga-channels)

**Comment by crizant** [comment]
Sure, it is not really solving the problem.
In real world application there should be actions being fired anytime. And any saga functions requests api could be causing this problem.
I think the middleware approach would be better.

**Comment by bmealhouse** [comment]
Yes, channel code is very tightly coupled to the use case. It has to be like this due to how `END` works. Let me know if you have any luck with the redux middleware approach. I will do some experimentation as well.

Have you considered any other async state management libraries that might be better suited for the application you are trying to build?

**Comment by crizant** [comment]
Actually my application is on production already, I am just trying to fix bugs.
It is too late to move to other libraries. =[

**Comment by bmealhouse** [comment]
@crizant, I did some experimentation and came up with a nice middleware to solve your problem.
[https://github.com/bmealhouse/next-redux-saga-issue/tree/redux-saga-middleware](https://github.com/bmealhouse/next-redux-saga-issue/tree/redux-saga-middleware)

I called it `record-replay-middleware`. All it does is record actions that have been dispatched after the saga has ended, and then, when the saga reruns, it replays the recorded actions.

Have a look when you get a chance and let me know if it works for you.

**Comment by crizant** [comment]
In this case, the actions will pass through reducers twice, right?
This may cause side effects.

**Comment by bmealhouse** [comment]
Oh duh, good point.

In that case, we need to use a custom `redux-saga` emitter instead.
{https://github.com/bmealhouse/next-redux-saga-issue/tree/redux-saga-emitter](https://github.com/bmealhouse/next-redux-saga-issue/tree/redux-saga-emitter)

It uses the same technique as before, but instead of re-dispatching actions, it only emits them to `redux-saga`.

**Comment by crizant** [comment]
It works perfectly! Thank you!
Could you please include the emitter in `next-redux-saga`?

**Comment by flavordaaave** [comment]
@bmealhouse Thanks for your suggested solutions using either a middleware or an emitter.
Unfortunately for me only the middleware approach is working. Basically I have 2 sagas. One to fetch a config file from remote and a second one that needs a URL from the config file to fetch some products. Any idea why I can't get the emitter solution working with the following basic setup?

```
// config-saga.js
function* loadConfig(action) {
  try {
    const result = yield call(loadConfigFromServer) 
    yield put({type: 'SET_CONFIG', payload: result})
  } catch (e) {
    yield put(configError(e))
  }
}

export default [
  takeLatest('LOAD_CONFIG', loadConfig),
]
```
```
// product-saga.js
function* fetchProducts(action) {
  try {
    // Ensure that the config is loaded
    const config = yield select(getConfig) // getConfig is a simple selector
    if (!config) yield take('SET_CONFIG')

    ... Call the API using the URL from the config
    yield put({type: 'FETCH_PRODUCTS_SUCCESS', payload: products})
  } catch (e) {
    yield put({type: 'FETCH_PRODUCTS_FAIL', payload: e})
  }
}

export default [
  takeLatest('FETCH_PRODUCTS_REQUEST', fetchProducts)
]
```
```
// pages/products.js
class Page extends Component {
  static async getInitialProps({ store, isServer, pathname, query }) {
    if (typeof window === 'undefined') {
      store.dispatch({type: 'LOAD_CONFIG'})
      store.dispatch({type: 'FETCH_PRODUCTS_REQUEST'})
    }
  }

  render() {
    // Render products using this.props.products
  }
}

function mapStateToProps(state) {
  return {
    products: getProducts(state), // getProducts is another simple selector
  }
}

export default connect(mapStateToProps)(Page)
```

Only when using the middleware solution the products are included in the store that is passed to the client.

**Comment by bmealhouse** [comment]
Sorry @flavordaaave, I haven't been able to get to your question yet. I don't use redux-saga anymore so it will take me a bit of time to get back up to speed with this. Unfortunately, I won't have time to check it out for another 2-3 weeks.

**Comment by flavordaaave** [comment]
@bmealhouse no worries. As mentioned I found a solution so no need for you to look into this at all. Thanks anyways.
